pkg/odo/cli/component/create_methods.go

```diff
@@ -0,0 +1,393 @@
+		Token: co.devfileMetadata.token,
+	}
+	devfileData, err := util.DownloadFileInMemory(params)
+	if err != nil {
+		return errors.Wrapf(err, "failed to download devfile for devfile component from %s", co.devfileMetadata.devfilePath.value)
+	}
+	err = ioutil.WriteFile(co.DevfilePath, devfileData, 0644) // #nosec G306
+	if err != nil {
+		return errors.Wrapf(err, "unable to save devfile to %s", co.DevfilePath)
+	}
+	devfileSpinner.End(true)
+
+	// SET METADATA
+	co.devfileMetadata.devfilePath.protocol = "http(s)"
+	co.devfileMetadata.componentName, co.devfileMetadata.componentType, err = getMetadataForExistingDevfile(co, args)
+	return err
+}
+
+func (hcm HTTPCreateMethod) Rollback(componentContext string) {
+	os.RemoveAll(componentContext)
+}
+
+// FileCreateMethod is used when --devfile flag is used with a local file; `odo create --devfile /tmp/comp/devfile.yaml`
+type FileCreateMethod struct{}
+
+func (fcm FileCreateMethod) CheckConflicts(co *CreateOptions, args []string) error {
+	return conflictCheckForDevfileFlag(args, co.devfileMetadata.devfileRegistry.Name)
+}
+
+func (fcm FileCreateMethod) FetchDevfileAndCreateComponent(co *CreateOptions, cmd *cobra.Command, args []string) error {
+	devfileAbsolutePath, err := filepath.Abs(co.devfileMetadata.devfilePath.value)
+	if err != nil {
+		return err
+	}
+	devfileSpinner := log.Spinnerf("Creating a devfile component from devfile path: %s", devfileAbsolutePath)
+	defer devfileSpinner.End(false)
+	devfileData, err := ioutil.ReadFile(co.devfileMetadata.devfilePath.value)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read devfile from %s", co.devfileMetadata.devfilePath)
+	}
+	err = ioutil.WriteFile(co.DevfilePath, devfileData, 0644) // #nosec G306
+	if err != nil {
+		return errors.Wrapf(err, "unable to save devfile to %s", co.DevfilePath)
+	}
+	devfileSpinner.End(true)
+
+	//	SET METADATA
+	co.devfileMetadata.devfilePath.protocol = "file"
+	co.devfileMetadata.componentName, co.devfileMetadata.componentType, err = getMetadataForExistingDevfile(co, args)
+	return err
+}
+
+func (fcm FileCreateMethod) Rollback(componentContext string) {
+	os.RemoveAll(componentContext)
+}
+
+// conflictCheckForDevfileFlag checks for the common conflicts while using --devfile flag
+func conflictCheckForDevfileFlag(args []string, registryName string) error {
+	// More than one arguments should not be allowed when --devfile is used
+	if len(args) > 1 {
+		return &DevfileExistsExtraArgsError{len(args)}
+	}
+	// Check if both --devfile and --registry flag are used, in which case raise an error
+	if registryName != "" {
+		return &DevfileFlagWithRegistryFlagError{}
+	}
+	return nil
+}
+
+// validateAndFetchRegistry validates if the provided registryName exists and returns the devfile listed in the registy;
+// if the registryName is "", then it returns devfiles of all the available registries
+func validateAndFetchRegistry(registryName string) (catalog.DevfileComponentTypeList, error) {
```

Comment:
Should we allow creating a component if no devfile registry is present for direct(`odo create <comp_name> <comp_type>`) and interactive mode?

@feloy @dharmit

Reply:
Yes, I think you could detect early that there is no registry configured and stop at this point

Reply:
Looks like this check has already been implemented.